### ISSUE TITLE
Index AEX-N primarly using events

### ISSUE THREAD

**Issue by thepiwo** [aeternity/ae_mdw#969]
As discussed we want to continuously index by tracking events, rather than dry-runs.

For AEX-9 we would need specifically one dry-run after contract creation as `init` won't give any events in the current implementation. Later account presence changes can be observed using Transfer and Mint Events. For balance updates I think we currently don't actively track them, single account balances could be dry-runned in constant time.

For AEX-141 I think we can have similar approach, but eventually skip dry-running at all (cc @marc0olo)

@jyeshe is there anything you would like to add. I think for planning on what specifically to do it would be nice to have a concise but full summary of the changes we want to do.

**Comment by jyeshe** [comment]
As contract samples after `init` state (on testnet):
`ct_ZwpiXbWme5EC72FCDoAvJMqkuwgGaJkKEyN5VCwewrYoWaTEW`, 3000 account balances, 1.2 second
`ct_HQyn8crD6APzS7VPguNKEdVMYqdupPG95rfQU2a9ztosFLtmh`, 4000 account balances 1.6 second

**Comment by thepiwo** [comment]
@jyeshe interesting so they created thousands of accounts in contract create?

**Comment by jyeshe** [comment]
> @jyeshe interesting so they created thousands of accounts in contract create?

Exactly

**Comment by jyeshe** [comment]
As discussed we gonna approach the contract `init` state with synchronous calls limiting the dry-run time with the `gas`.

**Comment by jyeshe** [comment]
Aex9 contracts with no logs returned by https://testnet.aeternity.art/mdw/v2/contracts/logs?contract=:contract_id

```
["ct_2cYkXZjXcd6qjnoZPPmH4NVpo6FR4PUYmA14mmckg7oUuqhDtj",
 "ct_NfTemzEwdWvLQ2wwjmpJuRjTHdTdqqtcqeVGmY9MhsbJHkk42",
 "ct_2mj7kpXoR36Y2iMrHFcZxww7fyzfmgZHNjqmmGUFoRHseVDgWF",
 "ct_HCQLprHjoTRZMb1kGDvStSa3bhK59b167SAzGBgDxuDCyD2pD",
 "ct_2Fj6Km8886DdRea4Tp92ggBaoAyknkMB1kG7o3CGux2CB1ynGC",
 "ct_2DbY6e5useaDjbd3bWX38e4tZYw7gS9ZgP79mbDFcgCmmYtaPq",
 "ct_2wL6ghE7SMwA5EGDeAV7A49SmfdSzaKNM8uLojPvrAhGgPsAYr",
 "ct_2R1bq5cdmG65jY5EXv7GX7X9BnccMa4Zv2XWjqvEbF7b4WxsJQ",
 "ct_hfsUEW55ZYt6fqfQyiq2oSQRSZvVijYDc25YA8aXUi2BHf2BR",
 "ct_PSAWd8sXhQXcUx6PirovLePboxgr8aioT4wpBnQBoTajXPHQ4"]
```

**Comment by jyeshe** [comment]
testnet aex9 contracts without logs:
https://testnet.aeternity.art/mdw/v2/contracts/logs?contract=ct_2cYkXZjXcd6qjnoZPPmH4NVpo6FR4PUYmA14mmckg7oUuqhDtj
https://testnet.aeternity.art/mdw/v2/contracts/logs?contract=ct_NfTemzEwdWvLQ2wwjmpJuRjTHdTdqqtcqeVGmY9MhsbJHkk42
https://testnet.aeternity.art/mdw/v2/contracts/logs?contract=ct_2mj7kpXoR36Y2iMrHFcZxww7fyzfmgZHNjqmmGUFoRHseVDgWF
https://testnet.aeternity.art/mdw/v2/contracts/logs?contract=ct_HCQLprHjoTRZMb1kGDvStSa3bhK59b167SAzGBgDxuDCyD2pD
https://testnet.aeternity.art/mdw/v2/contracts/logs?contract=ct_2Fj6Km8886DdRea4Tp92ggBaoAyknkMB1kG7o3CGux2CB1ynGC
https://testnet.aeternity.art/mdw/v2/contracts/logs?contract=ct_2DbY6e5useaDjbd3bWX38e4tZYw7gS9ZgP79mbDFcgCmmYtaPq
https://testnet.aeternity.art/mdw/v2/contracts/logs?contract=ct_2wL6ghE7SMwA5EGDeAV7A49SmfdSzaKNM8uLojPvrAhGgPsAYr
https://testnet.aeternity.art/mdw/v2/contracts/logs?contract=ct_2R1bq5cdmG65jY5EXv7GX7X9BnccMa4Zv2XWjqvEbF7b4WxsJQ
https://testnet.aeternity.art/mdw/v2/contracts/logs?contract=ct_hfsUEW55ZYt6fqfQyiq2oSQRSZvVijYDc25YA8aXUi2BHf2BR
https://testnet.aeternity.art/mdw/v2/contracts/logs?contract=ct_PSAWd8sXhQXcUx6PirovLePboxgr8aioT4wpBnQBoTajXPHQ4

**Comment by jyeshe** [comment]
testnet aex9 contracts with more new account balances (not present on contract creation) than logs:
https://testnet.aeternity.art/mdw/aex9/balances/gen/584147/ct_JDp175ruWd7mQggeHewSLS1PFXt9AzThCDaFedxon8mF8xTRF
https://testnet.aeternity.art/mdw/aex9/balances/gen/689551/ct_JDp175ruWd7mQggeHewSLS1PFXt9AzThCDaFedxon8mF8xTRF
https://testnet.aeternity.art/mdw/v2/contracts/logs?contract=ct_JDp175ruWd7mQggeHewSLS1PFXt9AzThCDaFedxon8mF8xTRF
38 new amounts for 10 events

https://testnet.aeternity.art/mdw/aex9/balances/gen/569487/ct_24gNuddxAbMYtT32sh8Xm1PpB2fZ3HMGtfST5sA3irect3Yu76
https://testnet.aeternity.art/mdw/aex9/balances/gen/689551/ct_24gNuddxAbMYtT32sh8Xm1PpB2fZ3HMGtfST5sA3irect3Yu76
https://testnet.aeternity.art/mdw/v2/contracts/logs?contract=ct_24gNuddxAbMYtT32sh8Xm1PpB2fZ3HMGtfST5sA3irect3Yu76
11 new amounts for 10 events

https://testnet.aeternity.art/mdw/aex9/balances/gen/269166/ct_2ftFoWeniE5Rgy6XwarqzW1F9Y8Pbwm9K6UmyXMQ1v1dPVxSHt
https://testnet.aeternity.art/mdw/aex9/balances/gen/689551/ct_2ftFoWeniE5Rgy6XwarqzW1F9Y8Pbwm9K6UmyXMQ1v1dPVxSHt
https://testnet.aeternity.art/mdw/v2/contracts/logs?contract=ct_2ftFoWeniE5Rgy6XwarqzW1F9Y8Pbwm9K6UmyXMQ1v1dPVxSHt
3 new amounts for 2 events

**Comment by jyeshe** [comment]
testnet aex9 contract with more new account balances (not present on contract creation) than logs:
https://testnet.aeternity.art/mdw/aex9/balances/gen/269166/ct_2ftFoWeniE5Rgy6XwarqzW1F9Y8Pbwm9K6UmyXMQ1v1dPVxSHt
https://testnet.aeternity.art/mdw/aex9/balances/gen/689551/ct_2ftFoWeniE5Rgy6XwarqzW1F9Y8Pbwm9K6UmyXMQ1v1dPVxSHt
https://testnet.aeternity.art/mdw/v2/contracts/logs?contract=ct_2ftFoWeniE5Rgy6XwarqzW1F9Y8Pbwm9K6UmyXMQ1v1dPVxSHt
3 new amounts for 2 events

**Comment by thepiwo** [comment]
> but with no balance after contract creation (non compliant with `Mint` or `Transfer` events):
> https://testnet.aeternity.art/mdw/aex9/balances/gen/343384/ct_2R1bq5cdmG65jY5EXv7GX7X9BnccMa4Zv2XWjqvEbF7b4WxsJQ

but for contract creation we still want to dry run due to the issues with AEX-9 reference implementation

**Comment by jyeshe** [comment]
These are contracts on similar situation where the event balance is missing but there are no event logs to sustain amount on the contract balances. The difference here is that there are some event logs but no aex9 event involving these accounts: 
```[
  {"ct_JDp175ruWd7mQggeHewSLS1PFXt9AzThCDaFedxon8mF8xTRF",
   "ak_Vdnb4zqWCAu5vKstsqhhJKd95tcJTCrEcJdvzaBKkf6Jz5dkE"},
  {"ct_JDp175ruWd7mQggeHewSLS1PFXt9AzThCDaFedxon8mF8xTRF",
   "ak_21YutbSDxVCC4y7mt2YDtCS3iqdrXSVCEH84PeyGTCDJUFnkfe"},
  {"ct_JDp175ruWd7mQggeHewSLS1PFXt9AzThCDaFedxon8mF8xTRF",
   "ak_2CQa7mKHsEdpq4ZGQ8bUk2y9bUiD9ZuibmcQjeCfma2YLFppyd"},
  {"ct_JDp175ruWd7mQggeHewSLS1PFXt9AzThCDaFedxon8mF8xTRF",
   "ak_2QwV57qAR1rPqWfX4smiTXTn6Gp3aRd2q7boGxJy74wEMn85N7"},
  {"ct_JDp175ruWd7mQggeHewSLS1PFXt9AzThCDaFedxon8mF8xTRF",
   "ak_2Wxcc2kF5Za5rVC2pzC9kUu6NzVmscstngTCpYZjc8eaCC1pg9"},
  {"ct_JDp175ruWd7mQggeHewSLS1PFXt9AzThCDaFedxon8mF8xTRF",
   "ak_2XYL7rNGhNEBgb7LwVzjUxFhZp67Dj2z7jqEBP7TmvFZTZiA15"},
  {"ct_JDp175ruWd7mQggeHewSLS1PFXt9AzThCDaFedxon8mF8xTRF",
   "ak_2bhuQQfUGfyzdBPiBZXbebBfBJ8YKHsAhvnnPdrYH8BQeYqn1G"},
  {"ct_JDp175ruWd7mQggeHewSLS1PFXt9AzThCDaFedxon8mF8xTRF",
   "ak_2kE1RxHzsRE4LxDFu6WKi35BwPvrEawBjNtV788Gje3yqADvwR"},
  {"ct_JDp175ruWd7mQggeHewSLS1PFXt9AzThCDaFedxon8mF8xTRF",
   "ak_2n5ttJFmskxQ7mB2CJbU6Roi8bFB5u33fBwnWJeF7Js25avtSj"},
  {"ct_JDp175ruWd7mQggeHewSLS1PFXt9AzThCDaFedxon8mF8xTRF",
   "ak_2nQcf5PRKWKiTJAjPJhnUDKc56CQ3pZcFrwYFYKEosqiunGuDV"},
  {"ct_JDp175ruWd7mQggeHewSLS1PFXt9AzThCDaFedxon8mF8xTRF",
   "ak_2naewT5mBXGUdYqdd79Tmkd4ba1EVDzxFk5FFWbFSGqYVdW156"},
  {"ct_24gNuddxAbMYtT32sh8Xm1PpB2fZ3HMGtfST5sA3irect3Yu76",
   "ak_2KXEqxKoAjbN1GrrggRatHhCySNL5zAzavUXqAAvjTH4kycFwx"},
  {"ct_2mdY71wG4zAjrdmqDJPXU6h8dYpzNs4mMZ81ujeNnrQPU2jMto",
   "ak_2AVeRypSdS4ZosdKWW1C4avWU4eeC2Yq7oP7guBGy8jkxdYVUy"},
  {"ct_2kc9naWGGnx4TWGK7UR9gut2cVcDvf7pv8CBYG1a8WML2jzUeb",
   "ak_sdgYaFPQuVeD9FrnCXFZD85wgDCoksd1nkTiD3p4ChQifqFUH"},
  {"ct_2kc9naWGGnx4TWGK7UR9gut2cVcDvf7pv8CBYG1a8WML2jzUeb",
   "ak_2AVeRypSdS4ZosdKWW1C4avWU4eeC2Yq7oP7guBGy8jkxdYVUy"}
]
```

**Comment by thepiwo** [comment]
> These are contracts on similar situation where the event balance is missing but there are no event logs to sustain certain amount on the contract balances. The difference here is that there are some event logs but no aex9 event involving these accounts:

summarising we are are all good with our new solution?

**Comment by jyeshe** [comment]
> > These are contracts on similar situation where the event balance is missing but there are no event logs to sustain certain amount on the contract balances. The difference here is that there are some event logs but no aex9 event involving these accounts:
> 
> summarising we are are all good with our new solution?

There are two different results for event and dry-run balances:
a) missing contract account balance
b) different amount for contract account balance
 
These above were all contracts for which there is at least one event balance missing (a) due to lack of an event related to the account with missing balance.
```
["ct_JDp175ruWd7mQggeHewSLS1PFXt9AzThCDaFedxon8mF8xTRF",
 "ct_24gNuddxAbMYtT32sh8Xm1PpB2fZ3HMGtfST5sA3irect3Yu76",
 "ct_2ftFoWeniE5Rgy6XwarqzW1F9Y8Pbwm9K6UmyXMQ1v1dPVxSHt",
 "ct_2mdY71wG4zAjrdmqDJPXU6h8dYpzNs4mMZ81ujeNnrQPU2jMto",
 "ct_2kc9naWGGnx4TWGK7UR9gut2cVcDvf7pv8CBYG1a8WML2jzUeb"]
 ```
This one for having no log: 
https://testnet.aeternity.art/mdw/v2/contracts/logs?contract=ct_2R1bq5cdmG65jY5EXv7GX7X9BnccMa4Zv2XWjqvEbF7b4WxsJQ

For different amounts, there are these ones that will be analyzed:
["ct_wGJDBU7bb6RyuDVkZuZCkbtMZmzsftGCNpS9FXZB84s16ymzG",
 "ct_qACQJ9o8pfywFg11c8Au47vyPK5qTNzS9g4DEnB2j7FPQV9tW",
 "ct_BWm6c3FAXzd5YUWodHGH7bSpYQHzgkNccDoR7uRSSgoUDTKvB"]

**Comment by jyeshe** [comment]
For both `ct_bVg6KvgniPBuE8N5KmjvhH4P23kkFGJe1yqiadxLgtyLEYmtL` accounts there is a transfer event that changes the balance so even though the balance is not the one stored in the contract it's correct according to the contract creation and event amounts.

For `ct_2scZsZ2Us2iaRTox4kRUezBiE8C97ZeXedM1mXne6TeFuN5bt2` there is an event on contract creation that was computed but should not. Fixed by: https://github.com/aeternity/ae_mdw/pull/1020

**Comment by jyeshe** [comment]
@thepiwo after #1020 to close the issue it will be pending the improvement on contract creation dry-run error handling (for a first version it depends on a dry-run with small timeout that might fail and this timeout cannot be increased with the risk of blocking the syncing of all other transactions).

**Comment by thepiwo** [comment]
Thanks for the thorough investigation, first of all very nice that we just have these few we are unsure about and I guess we will find some logical explaination

**Comment by jyeshe** [comment]
> Thanks for the thorough investigation, first of all very nice that we just have these few we are unsure about and I guess we will find some logical explanation.
> 
> > For both `ct_bVg6KvgniPBuE8N5KmjvhH4P23kkFGJe1yqiadxLgtyLEYmtL` accounts there is a transfer event that changes the balance so even though the balance is not the one stored in the contract it's correct according to the contract creation and event amounts.
> 
> so this one we assume they somehow messed up their contract source and give a transfer event where no transfer was happening? Can it be that there are multiple transfers in one transaction, or even remote contract calls?

Yes, I was surprised with so many compliant contracts (with real balances matching event balances).
For this contract there is some discrepancy. Multiple transfer events and remote call events are handled by current implementation (both using a different log index for each event). Pushed now more thorough test coverage though.

**Comment by thepiwo** [comment]
awesome, thanks for the feedback!
